# 1. Introdução

O trabalho de um analista é analisar dados para obter percepções valiosas dos dados e tomar decisões fundamentadas neles. Esse processo consiste em várias etapas, como visão geral dos dados, pré-processamento dos dados e testes de hipóteses.

Sempre que fazemos uma pesquisa, precisamos formular uma hipótese que depois poderemos testar. Às vezes nós aceitamos essas hipóteses; outras vezes, nós as rejeitamos. Para fazer as escolhas certas, um negócio deve ser capaz de entender se está fazendo as suposições certas ou não.

Neste projeto, você vai comparar as preferências musicais dos habitantes de Springfild e Shelbyville. Você vai estudar os dados de um serviço de streaming de música online para testar a hipótese apresentada abaixo e comparar o comportamento dos usuários dessas duas cidades.

### 1.1 Objetivo:

Teste a hipótese:

1. A atividade dos usuários é diferente dependendo do dia da semana e da cidade.

### 1.2 Etapas:

Os dados sobre o comportamento do usuário são armazenados no arquivo /datasets/music_project_en.csv. Não há informações sobre a qualidade dos dados, então será necessário examiná-los antes de testar a hipótese.

Primeiro, você avaliará a qualidade dos dados e verá se seus problemas são significativos. Depois, durante o pré-processamento dos dados, você tentará tratar dos problemas mais críticos.

O seu projeto consistirá em três etapas:

1. Visão geral dos dados
2. Pré-processamento de dados
3. Teste da hipótese


# 2. Etapa 1: Visão geral dos dados

Abra os dados e examine-os.

Você precisará da pandas, então, importe-a.

In [1]:
# importação de biblioteca

import pandas as pd

Leia o arquivo music_project_en.csv da pasta /datasets/ e salve-o na variável df:

In [2]:
# Carregamento de dados

df = pd.read_csv('music_project_en.csv')

Imprima as primeiras 10 linhas da tabela:

In [3]:
# obtenha as 10 primeiras 10 linhas da tabela df

print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

Obtenha informações gerais sobre a tabela usando um comando. Você conhece o método para exibir informações gerais que precisamos obter.

In [4]:
# obtendo informações gerais sobre os nossos dados

print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


Aqui estão as nossas observações sobre a tabela. Ela contém sete colunas. Elas armazenam o mesmo tipo de dado: object.

De acordo com a documentação:

- ' userID' — identificação do usuário
- 'Track' — título da música
- 'artist' — nome do artista
- 'genre' — gênero da música
- 'City' — cidade do usuário
- 'time' — o tempo exato que a música foi reproduzida
- 'Day' — dia da semana

Podemos ver três problemas de estilo nos cabeçalhos da tabela:

1. Alguns cabeçalhos são escritos em letras maiúsculas, outros estão em minúsculas.
2. Alguns cabeçalhos contêm espaços.
3. Detecte o problema e o descreva aqui.


Análise dos cabeçalhos: ERROS: 
1. Excesso de espaços;
2. Mistura no padrão de escritas (maiusculas e minusculas);
3. Apesar de ter apenas um dos cabeçalhos com mais de uma palavra de identificação (user id), faz-se necessário o uso do caractere de separação '_'.

SOLUÇÕES:

Para solucionar os erros identificados usaremos os métodos:
- .strip() - solução para os espaçoes em excesso 
- .lower() - para igualar todas a letras em minusculas 
- .replace() - para substituir os espaços(' ') entre as palavras por underlines ('_')


### 2.1 Escreva suas observações. Aqui estão algumas perguntas que podem lhe ajudar

1.   Que tipo de dados temos nas linhas? 

E como podemos entender as colunas? Os dados são do tipo object. Apesar dos erros como espaçamentos em excesso, mistura de letras maiusculas e minuscula, e falta de separação das palavras, conseguiremos trablhar com os dados disponibilizados. Claro que depois de tratar os dados das colunas.

2.   Esses dados são suficientes para responder à nossa hipótese ou precisamos de mais dados?

Depois dos tratamentos acreditamos que os dados podem sim ajudar na decisão e responder as hipóteses.

3.   Você notou algum problema nos dados, como valores ausentes, duplicados ou tipos de dados errados?

Depois de uma análise prévia nota-se que a tabela possui erros de conteúdos, como valos nulos e ausentes.

# 3. Etapa 2: Pré-processamento de dados

O objetivo aqui é preparar os dados para a análise. O primeiro passo é resolver todos os problemas com o cabeçalho. E então podemos passar para os valores ausentes e duplicados. Vamos começar.

Corrija a formatação nos cabeçalhos da tabela.

### 3.1 Estilo do cabeçalho

Imprima os cabeçalhos da tabela (os nomes das colunas):

In [6]:
# imprima os nomes das colunas

print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Mude os cabeçalhos da tabela conforme as boas práticas de estilo:

- Todos os caracteres precisam estar com letras minúsculas
- Exclua espaços
- Se o nome tiver várias palavras, use snake_case

Anteriormente, você aprendeu sobre uma maneira automatizada de renomear colunas. Vamos usá-la agora. Use o ciclo for para percorrer os nomes das colunas e transformar todos os caracteres em letras minúsculas. Após fazer isso, imprima os cabeçalhos da tabela novamente:



In [7]:
# Percorrendo os cabeçalhos e convertendo tudo em minúsculos

new_col = []
for linha in df.columns:
    col_minusculas = linha.lower()
    new_col.append(col_minusculas)

print(new_col)

['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day']


Agora, usando a mesma abordagem, exclua os espaços no início e no final de cada nome de coluna e imprima os nomes das colunas novamente:

In [8]:
# Percorrendo os cabeçalhos e removendo os espaços

new_col = []
for linha in df.columns:
    col_minusculas = linha.lower()
    col_striped = col_minusculas.strip()
    new_col.append(col_striped)

print(new_col)

['userid', 'track', 'artist', 'genre', 'city', 'time', 'day']


Precisamos aplicar a regra de sublinhado no lugar de espaço à coluna userid. Deveria ser user_id. Renomeie essa coluna e imprima os nomes de todas as colunas quando terminar.

In [9]:
# Renomeando a coluna "userid"

new_col = []
for linha in df.columns:
    col_minusculas = linha.lower()
    col_striped = col_minusculas.strip()
    col_no_space = col_striped.replace('userid', 'user_id')
    new_col.append(col_no_space)

print(new_col)

['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day']


Verifique o resultado. Imprima os cabeçalhos novamente:

In [10]:
# verificando o resultado: a lista de cabeçalhos

new_col = []
for linha in df.columns:
    col_minusculas = linha.lower()
    col_striped = col_minusculas.strip()
    col_no_space = col_striped.replace('userid', 'user_id')
    new_col.append(col_no_space)

df.columns = new_col
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


### 3.2 Valores ausentes

Primeiro, encontre a quantidade de valores ausentes na tabela. Você precisa usar dois métodos em sequência para obter o número de valores ausentes.

In [11]:
# calculando o número de valores ausentes

print(df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


Nem todos os valores ausentes afetam a pesquisa. Por exemplo, os valores ausentes em track e artist não são críticos. Você pode simplesmente substituí-los por valores padrão, como a string 'unknown'.

Mas valores ausentes em 'genre' podem afetar a comparação de preferências musicais de Springfield e Shelbyville. Na vida real, seria útil descobrir as razões pelas quais os dados estão ausentes e tentar corrigi-los. Mas nós não temos essa possibilidade neste projeto. Então, você terá que:

- Preencha esses valores ausentes com um valor padrão
- Avalie em que medida os valores ausentes podem afetar sua análise


Substitua os valores ausentes nas colunas 'track', 'artist' e 'genre' pela string 'unknown'. Como mostramos nas lições anteriores, a melhor maneira de fazer isso é criar uma lista para armazenar os nomes das colunas nas quais precisamos fazer a substituição. Em seguida, use essa lista e percorra as colunas nas quais a substituição seja necessária e faça a substituição.

In [13]:
# percorrendo os cabeçalhos e substituindo valores ausentes por 'unknown'

df['track'].fillna('unknown', inplace=True)
df['artist'].fillna('unknown', inplace=True)
df['genre'].fillna('unknown', inplace=True)

print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


Agora verifique o resultado para ter certeza de que o conjunto de dados não contenha valores ausentes após a substituição. Para fazer isso, conte os valores ausentes novamente.

In [14]:
# contando os valores ausentes

print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


### 3.3 Valores duplicados

Encontre o número de duplicados explícitos na tabela. Lembre-se de que você precisa aplicar dois métodos em sequência para obter o número de duplicados explícitos.

In [15]:
# contando duplicados explícitos

print(df.duplicated().sum())

3826


Agora descarte todos os duplicados. Para fazer isso, chame o método que faz exatamente isso.

In [16]:
# removendo duplicados explícitos

df.drop_duplicates(inplace = True)


Agora vamos verificar se descartamos todos os duplicados. Conte duplicados explícitos mais uma vez para ter certeza de que você removeu todos eles:

In [17]:
# verificando duplicados novamente

print(df.duplicated().sum())

0


Agora queremos nos livrar dos duplicados implícitos na coluna genre. Por exemplo, o nome de um gênero pode ser escrito de maneiras diferentes. Alguns erros afetarão também o resultado.

Para fazer isso, vamos começar imprimindo uma lista de nomes de gênero únicos, ordenados em ordem alfabética: Para fazer isso:

- Extraia a coluna genre do DataFrame
- Chame o método que retornará todos os valores únicos na coluna extraída


In [19]:
# visualizando nomes de gêneros únicos

print(df['genre'].unique())

['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hip' 'jazz' 'postrock' 'latin'
 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental' 'rusrock'
 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'hiphop' 'drum' 'extrememetal' 'türkçe'
 'experimental' 'easy' 'metalcore' 'modern' 'argentinetango' 'old' 'swing'
 'breaks' 'eurofolk' 

Olhe a lista e encontre duplicados implícitos do gênero hiphop. Esses podem ser nomes escritos incorretamente, ou nomes alternativos para o mesmo gênero.

Você verá os seguintes duplicados implícitos:

- hip
- hop
- hip-hop

Para se livrar deles, crie uma função replace_wrong_genres() com dois parâmetros:

- wrong_genres= — essa é uma lista que contém todos os valores que você precisa substituir
- correct_genre= — essa é uma string que você vai usar para a substituição

Como resultado, a função deve corrigir os nomes na coluna 'genre' da tabela df, isto é, substituindo cada valor da lista wrong_genres por valores de correct_genre.

Dentro do corpo da função, use um ciclo 'for' para percorrer a lista de gêneros errados, extrair a coluna 'genre' e aplicar o método replace para fazer as correções.

In [20]:
# função para substituir duplicados implícitos

def replace_wrong_genres(df, column, wrong_genres, correct_genre):
    df[column] = df[column].replace(wrong_genres, correct_genre)
    return df

Agora, chame a função replace_wrong_genres() e passe argumentos apropriados para que ela limpe duplicados implícitos (hip, hop e hip-hop) substituindo-os por hiphop:

In [21]:
# removendo duplicados implícitos

duplicates = ['hip', 'hiphop']
name = 'hiphop'
df = replace_wrong_genres(df, 'genre', duplicates, name)

duplicates = ['hop', 'hiphop']
name = 'hiphop'
df = replace_wrong_genres(df, 'genre', duplicates, name)

duplicates = ['hip-hop', 'hiphop']
name = 'hiphop'
df = replace_wrong_genres(df, 'genre', duplicates, name)

print(df)

        user_id                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina           unknown   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfi

Certifique-se que os nomes duplicados foram removidos. Imprima a lista de valores únicos da coluna 'genre' mais uma vez:

In [22]:
# verificando valores duplicados

print(df['genre'].unique())

['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hiphop' 'jazz' 'postrock'
 'latin' 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental'
 'rusrock' 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'drum' 'extrememetal' 'türkçe' 'experimental' 'easy'
 'metalcore' 'modern' 'argentinetango' 'old' 'swing' 'breaks' 'eurofolk'
 'stone

### 3.4 Suas Observações

Descreva brevemente o que você reparou ao analisar duplicados, bem como a abordagem que usou para eliminá-los e os resultados que alcançou.

Primeiro ponto que chamou a atenção foi a quantidade de dados duplicados, foram encontrado 3826 dados duplicados explicitos, quantidade relevante que poderia interferir no resultado final da análise. O uso da função replace_wrong_genres(), foi de fundamental importancia, com ela o código ficou mais limpo e de fácil compreensão.

# 4. Etapa 3: Teste da hipótese

### 4.1. Hipótese: comparação do comportamento dos usuários nas duas cidades

A hipótese afirma que existem diferenças no consumo de música pelos usuários em Springfield e em Shelbyville. Para testar a hipótese, use os dados dos três dias da semana: segunda-feira (Monday), quarta-feira (Wednesday) e sexta-feira (Friday).

- Agrupe os usuários por cidade.
- Compare o número de músicas tocadas por cada grupo na segunda, quarta e sexta.

Execute cada cálculo separadamente.

O primeiro passo é avaliar a atividade dos usuários em cada cidade. Não se esqueça das etapas "divisão-aplicação-combinação" sobre as quais falamos anteriormente na lição. Agora seu objetivo é agrupar os dados por cidade, aplicar o método de contagem apropriado durante a etapa de aplicação e então encontrar o número de músicas tocadas por cada grupo, especificando a coluna para a qual você quer obter a contagem.

Veja um exemplo de como o resultado final deve ser: df.groupby(by='....')['column'].method() Execute cada cálculo separadamente.

Para avaliar a atividade dos usuários em cada cidade, agrupe os dados por cidade e encontre o número de músicas reproduzidas em cada grupo.



In [23]:
# Contando as músicas tocadas em cada cidade

print(df.groupby(by='city').count())

             user_id  track  artist  genre   time    day
city                                                    
Shelbyville    18512  18512   18512  18512  18512  18512
Springfield    42741  42741   42741  42741  42741  42741


Comente sobre suas observações aqui

A cidade de Springfield possui a quase 71% do ouvintes total.

Agora vamos agrupar os dados por dia da semana e encontrar a quantidade de músicas tocadas na segunda, quarta e sexta-feira. Use a mesma abordagem que antes, mas agora precisamos agrupar os dados de uma forma diferente.



In [24]:
# Calculando as músicas escutadas em cada um desses três dias

print(df.groupby(by='day').count())

           user_id  track  artist  genre   city   time
day                                                   
Friday       21840  21840   21840  21840  21840  21840
Monday       21354  21354   21354  21354  21354  21354
Wednesday    18059  18059   18059  18059  18059  18059


Comente sobre suas observações aqui Percebemos que ah um certo equilibrios entre os dias sexta e segunda. A quarta possui uma quantidade um pouco inferior em relação aos outros dois dias.

Você acabou de aprender como contar entradas agrupando-as por cidade ou por dia. E agora você precisa escrever uma função que possa contar entradas simultaneamente com base em ambos os critérios.

Crie a função number_tracks() para calcular o número de músicas tocadas em um determinado dia e em uma determinada cidade. A função deve aceitar dois parâmetros:

- day: um dia da semana pelo qual precisamos filtrar os dados. Por exemplo, 'Monday'.
- city: uma cidade pela qual precisamos filtrar os dados. Por exemplo, 'Springfield'.

Dentro da função, você vai aplicar uma filtragem consecutiva com indexação lógica.

Primeiro, filtre os dados por dia e então filtre a tabela resultante por cidade.

Depois de filtrar os dados usando os dois critérios, conte o número de valores na coluna 'user_id' da tabela resultante. O resultado da contagem representará o número de entradas que você quer encontrar. Armazene o resultado em uma nova variável e imprima-o.



In [25]:
# Declare a função number_tracks() com dois parâmetros: day= e city=.
def number_tracks(day, city):
  #  dia = day
  #  citie = city

  #  total_ouvintes = df[df['day'] == day]['user_id'].count() & df[df['city'] == city]['user_id'].count()
    
   # return print('No dia {} e na cidade {} tiveram {} ouvintes.'.format(dia, citie, total_ouvintes))

    # Armazene as linhas do DataFrame em que o valor na coluna 'day' é igual ao parâmetro day=
    filtro_dia = df[df['day'] == day]
    # Filtre as linhas em que o valor na coluna 'city' é igual ao parâmetro city=
    filtro_city = filtro_dia[filtro_dia['city'] == city]
    # Extraia a coluna 'user_id' da tabela filtrada e aplique o método count()
    contar_user_id = filtro_city['user_id'].count()
    # Retorne o número dos valores da coluna 'user_id'
    return print('No dia {} e na cidade {} tiveram {} ouvintes.'.format(day, city, contar_user_id))

Chame a função number_tracks() seis vezes, mudando os valores dos parâmetros, para que você possa recuperar os dados de ambas as cidades para cada um dos três dias.

In [26]:
# a quantidade de músicas tocadas em Springfield na segunda-feira

print(number_tracks('Monday', 'Springfield'))

No dia Monday e na cidade Springfield tiveram 15740 ouvintes.
None


In [27]:
# a quantidade de músicas tocadas em Shelbyville na segunda-feira

print(number_tracks('Monday', 'Shelbyville'))

No dia Monday e na cidade Shelbyville tiveram 5614 ouvintes.
None


In [28]:
# a quantidade de músicas tocadas em Springfield na quarta-feira

print(number_tracks('Wednesday', 'Springfield'))

No dia Wednesday e na cidade Springfield tiveram 11056 ouvintes.
None


In [29]:
# a quantidade de músicas tocadas em Shelbyville na quarta-feira

print(number_tracks('Wednesday', 'Shelbyville'))

No dia Wednesday e na cidade Shelbyville tiveram 7003 ouvintes.
None


In [30]:
# a quantidade de músicas tocadas em Springfield na sexta-feira

print(number_tracks('Friday', 'Springfield'))

No dia Friday e na cidade Springfield tiveram 15945 ouvintes.
None


In [31]:
# a quantidade de músicas tocadas em Shelbyville na sexta-feira

print(number_tracks('Friday', 'Shelbyville'))

No dia Friday e na cidade Shelbyville tiveram 5895 ouvintes.
None


# Conclusões

Comente sobre se a terceira hipótese está correta ou deve ser rejeitada. Explique seu raciocínio.

Com estes dados podemos confirmar que os habitantes de Springfield escutam mais musicas nas sextas e domingo. Enquanto os ouvintes de Shelbyville se concentram na quanta feira.
Contudo podemos dizer que os habitantes de Springfield mantiveram o percentual em torno de 70% nas sextas e domingo, tendo uma pequena queda para em torno de 60% na quarta-feira.

Concluimos que existe mais ouvintes na cidade de Springfield. Possuindo uma grande concentração nas sextas e domingos.
